include/Common/SharedQueue.h

```diff
@@ -0,0 +1,90 @@
+#pragma once
+
+#include <queue>
+#include <mutex>
+#include <condition_variable>
+
+template <typename T>
+class SharedQueue
```

Comment:
Why is this class even needed? Does [ConcurrentQueue](https://github.com/GrinPlusPlus/GrinPlusPlus/blob/master/include/Common/ConcurrentQueue.h) not do what you need?

Reply:
that class is subject to race conditions between calls to empty, front and pop if there is more than one thread; we could either use SharedQueue class or edit the ConcurrentQueue class. More here https://www.justsoftwaresolutions.co.uk/threading/implementing-a-thread-safe-queue-using-condition-variables.html